lib/api/BackbeatAPI.js

```diff
@@ -214,57 +214,85 @@ class BackbeatAPI {
-            const value = JSON.parse(JSON.parse(entry).value);
-            response.Versions.push({
-                Bucket: bucket,
-                Key: key,
-                VersionId: versionId,
-                StorageClass: site,
-                Size: value['content-length'],
-                LastModified: value['last-modified'],
-            });
-        }
-        return response;
+        const cmds = keys.map(k => ['get', k]);
+        return this._redisClient.batch(cmds, (err, res) => {
+            if (err) {
+                return cb(err);
+            }
+            for (let i = 0; i < res.length; i++) {
+                const [cmdErr, value] = res[i];
+                if (cmdErr) {
+                    return cb(cmdErr);
+                }
+                const queueEntry = QueueEntry.createFromKafkaEntry({ value });
+                response.Versions.push({
+                    Bucket: queueEntry.getBucket(),
+                    Key: queueEntry.getObjectKey(),
+                    VersionId: queueEntry.getEncodedVersionId(),
+                    StorageClass: queueEntry.getSite(),
+                    Size: queueEntry.getContentLength(),
+                    LastModified: queueEntry.getLastModified(),
+                });
+            }
+            return cb(null, response);
+        });
     }
 
     /**
-     * Find all failed CRR operations that match the bucket, key, and versionID.
-     * @param {Object} details - The route details
-     * @param {Function} cb - The callback to call
+     * Recursively scan all existing keys with a count of 1000. Call callback if
+     * the response is greater or equal to 1000 keys, or we have scanned all
+     * keys (i.e. when the cursor is 0).
+     * @param {String} pattern - The key pattern to match
+     * @param {Number} marker - The cursor to start scanning from
+     * @param {Array} allKeys - The collection of all matching keys found
+     * @param {Function} cb - The callback function
      * @return {undefined}
      */
-    getFailedCRR(details, cb) {
-        const { bucket, key, versionId } = details;
-        const pattern = `${bucket}:${key}:${versionId}:*`;
-        const cmds =
-            ['hscan', redisKeys.failedCRR, 0, 'MATCH', pattern, 'COUNT', 1000];
-        this._redisClient.batch([cmds], (err, res) => {
+    _scanAllKeys(pattern, marker, allKeys, cb) {
+        const cmd = ['scan', marker, 'MATCH', pattern, 'COUNT', 1000];
+        this._redisClient.batch([cmd], (err, res) => {
             if (err) {
                 return cb(err);
             }
             const [cmdErr, collection] = res[0];
             if (cmdErr) {
                 return cb(cmdErr);
             }
-            const [cursor, hashes] = collection;
-            return cb(null, this._getFailedCRRResponse(cursor, hashes));
+            const [cursor, keys] = collection;
+            allKeys.push(...keys);
+            if (allKeys.length >= 1000 || Number.parseInt(cursor, 10) === 0) {
```

Comment:
`allKeys.length >= 1000` 
If we get 1000 keys initially, won't this return before getting next set of keys? Or was this intentional?

Reply:
Yes, in effect it allows for a paginated response. I wanted to limit the API response listing to ~1000 keys to model closer to the default for version listings in S3. We cannot guarantee the exact number due to many keys Redis returns during the scan so if it's 1000 or more, the API will include a `NextMarker` value for subsequent listings.